*** manipulate.pm	Mon Apr 20 14:52:37 1998
--- /home/jesse/hyatt-manipulate.pm	Thu May 21 14:29:45 1998
***************
*** 37,70 ****
      
      $AuthRealm="WebRT for $rt::rtname";
      
-     
      ($name, $pass)=&WebAuth::AuthCheck($AuthRealm);
      
!     #if the user's password is bad
!     if (!(&rt::is_password($name, $pass))) {
!       
!       &WebAuth::AuthForceLogout($AuthRealm);
        exit(0);
      }
      
!     #if the user isn't even authenticating
!     elsif ($name eq '') {
        &WebAuth::AuthForceLogin($AuthRealm);
        exit(0)
      }
      
!     #if the user is trying to log out
!     if ($rt::ui::web::FORM{'display'} eq 'Logout') {
!       &WebAuth::AuthForceLogin($AuthRealm);
!       exit(0);
!     }
!     else {
        $current_user = $name;
        &WebAuth::Headers_Authenticated();
      }
      
      
  }
  sub InitDisplay {
    
    if ($rt::ui::web::FORM{'display'} eq 'SetNotify') { # #this is an ugly hack, but to get the <select>
--- 37,84 ----
      
      $AuthRealm="WebRT for $rt::rtname";
  
      ($name, $pass)=&WebAuth::AuthCheck($AuthRealm);
      
!     #if the user is trying to log out
!     if ($rt::ui::web::FORM{'display'} eq 'Logout') {
!       &WebAuth::AuthForceLogin($AuthRealm);
        exit(0);
      }
  
!     if ($name eq '') {
        &WebAuth::AuthForceLogin($AuthRealm);
        exit(0)
      } 
  
!     if (&rt::is_password($name, $pass)) {
!       # User is a RT user
        $current_user = $name;
+       $NS_Authenticated = 0;
        &WebAuth::Headers_Authenticated();
+     } elsif (ns_is_password($name, $pass)) {
+       # User has a local account
+       $current_user 			 = "guest";
+       $q_user_other 			 = $name;
+       $rt::ui::web::FORM{'q_user'}       = "other";
+       $rt::ui::web::FORM{'q_user_other'} = $name;
+       $NS_Authenticated = 1;
+       &WebAuth::Headers_Authenticated();
+     } else {
+       # Good bye
+       &WebAuth::AuthForceLogout($AuthRealm);
+       exit(0);
+     }
  }
  
+ sub ns_is_password($name, $pass) {
+     my($pass2);
+     my($name2, $encoded) = getpwnam($name);
  
+     $pass2    = crypt($pass, $encoded);
+     if ($pass eq $pass2) { return 1; }
+     return 0;
  }
+ 
  sub InitDisplay {
    
    if ($rt::ui::web::FORM{'display'} eq 'SetNotify') { # #this is an ugly hack, but to get the <select>
***************
*** 166,175 ****
--- 180,193 ----
      }
      
      elsif ($rt::ui::web::FORM{'display'} eq 'Create') {
+       if ($NS_Authenticated) { $current_user = $rt::ui::web::FORM{'q_user_other'}; }
        &FormCreate();
+       if ($NS_Authenticated) { $current_user = "guest"; }
      }
      elsif ($rt::ui::web::FORM{'display'} eq 'Create_Step2') {
+       if ($NS_Authenticated) { $current_user = $rt::ui::web::FORM{'q_user_other'}; }
        &FormCreate_Step2();
+       if ($NS_Authenticated) { $current_user = "guest"; }
      }
      
      elsif ($rt::ui::web::FORM{'display'} eq 'ShowNum') {
***************
*** 221,244 ****
    }
    
    if ($rt::ui::web::FORM{'display'} eq 'History') {
!     	  if (!$frames) {
! 	      &display_commands();
! 	        }
!     &do_bar($serial_num);
      
      &display_summary($serial_num);
      print "<hr>";
      &display_history_tables($serial_num);
      &do_bar($serial_num);
- 
- 	
    }
    
-   
-   
-   if (!$frames) {
-     &display_commands();
    }	
    &rt::ui::web::footer(); 
  }
  
--- 239,266 ----
    }
    
    if ($rt::ui::web::FORM{'display'} eq 'History') {
!     	  if (!$frames) { &display_commands(); }
  
+     $rt::req[$serial_num]{'requestors'} =~ /(.+)\@.+/;
+     if (($NS_Authenticated) && 
+         ( ($rt::req[$serial_num]{'requestors'} ne $rt::ui::web::FORM{'q_user_other'}) &&
+           ($1                                  ne $rt::ui::web::FORM{'q_user_other'})
+         )) {
+ 	  print "<BR><BR><BR><CENTER><H1> You can only view transactions you own. </H1></CENTER>\n";
+ 	  print "Owner $rt::req[$serial_num]{'requestors'}  OR $1    <BR>\n";
+           print "User  $rt::ui::web::FORM{'q_user_other'} <BR>   \n";
+           print "SN $serial_num \n";
+     } else {
+  	  &do_bar($serial_num);
  	  &display_summary($serial_num);
  	  print "<hr>";
   	  &display_history_tables($serial_num);
  	  &do_bar($serial_num);
      }
  	
    } 
+   
+   if (!$frames) { &display_commands(); }	
    &rt::ui::web::footer(); 
  }
  
